### PR TITLE
Pull mode

### PR DESCRIPTION
**Pull request by DIBL**

### CODE REVIEW

File: src/main/java/com/Elessar/app/server/DirectMsgSender.java [hunk 1]
```diff
@@ -16,14 +18,19 @@
...
 
-    public DirectMsgSender(HttpClient httpClient) {
+    public DirectMsgSender(HttpClient httpClient, MetricManager metricManager) {
         this.httpClient = httpClient;
+        this.metricManager = metricManager;
     }
 
     @Override
     public P2PMsgResponse send(List<Message> messages, String URL) throws Exception {
+        final Metric metric = metricManager.newMetric(new StringBuilder().append(MsgSender.MSG_SENDER).append(".")
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/server/DirectMsgSender.java]: Should we change the metric name to something like "DirectMsgSender.send"?

File: src/main/java/com/Elessar/app/server/KafkaMsgSender.java [hunk 2]
```diff
@@ -40,14 +47,21 @@ public P2PMsgResponse send(List<Message> messages, String topic) throws Exceptio
...
 
+        topic = topic.replaceAll("\\s", "");
         final ProducerRecord<Long, String> record = new ProducerRecord<>(topic, p2pMsgRequest.toString());
-        final RecordMetadata metadata = producer.send(record).get();
+        final P2PMsgResponse.Builder p2pMsgResponse = P2PMsgResponse.newBuilder();
+        final RecordMetadata metadata = producer.send(record).get();    // Synchronous send and wait for ack before proceed
 
-        logger.trace("Message sent to Kafka server at topic " + topic + " partition " + metadata.partition() + " with offset " + metadata.offset());
+        logger.trace("Message sent to Kafka server at topic " + metadata.topic() + " partition " + metadata.partition() + " with offset " + metadata.offset());
+        p2pMsgResponse.setSuccess(true).setIsDelivered(true);
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/server/KafkaMsgSender.java]: Please add a comment here explaining why we set the delivered flag to true, for example what the assumptions we make and which cases we don't handle and so on as this is an important design decision.

File: src/main/java/com/Elessar/app/server/MyServer.java [hunk 3]
```diff
@@ -38,7 +38,7 @@ public MyServer(String serverName, int port, MyDatabase db, LoadingCache<String,
         this.port = port;
         this.db = db;
         this.httpClient = new HttpClient(new NetHttpTransport().createRequestFactory());
-        this.msgSender = new DirectMsgSender(httpClient);
+        this.msgSender = new KafkaMsgSender(metricManager);
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/server/MyServer.java]: Same as what I mentioned for ClientMain, here we should pass in a flag indicating pull vs. push and create sender based on the flag.

File: src/main/java/com/Elessar/app/util/kafka/ProducerCreator.java [hunk 4]
```diff
@@ -14,6 +14,9 @@
         Properties props = new Properties();
         props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
         props.put(ProducerConfig.CLIENT_ID_CONFIG, "localhost:9092");
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/util/kafka/ProducerCreator.java]: This should be MyServer's url/port

File: src/main/java/com/Elessar/app/util/kafka/ProducerCreator.java [hunk 5]
```diff
@@ -14,6 +14,9 @@
...
+        props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);  // exactly once delivery => require in-flight requests < 5, retries > 0 and ack = all
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/util/kafka/ProducerCreator.java]: Can you provide a link to where those requirements come from?

File: src/main/java/com/Elessar/app/server/LogOnHandler.java [hunk 6]
```diff
@@ -93,7 +93,7 @@ public void handle(HttpExchange he) throws IOException {
 
                 // Get the list of unread messages sent to user
                 List<Message> messages = db.find(new Message(null, userName, null, null, false));
-                sendMessages(messages, clientURL);
+                sendMessages(messages, userName);
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/server/LogOnHandler.java]: pass the pull vs. push flag to LogOnHandler and use that to decide whether to pass userName or clientURL

File: src/main/java/com/Elessar/app/server/DirectMsgSender.java (hunk 1 above)
Comment (DIBL) [reply src/main/java/com/Elessar/app/server/DirectMsgSender.java]: I thought of this before, but in database, it seems we didn't differentiate this metric name between different database

File: src/main/java/com/Elessar/app/util/kafka/ProducerCreator.java (hunk 5 above)
Comment (DIBL) [reply src/main/java/com/Elessar/app/util/kafka/ProducerCreator.java]: >enable.idempotence
When set to 'true', the producer will ensure that exactly one copy of each message is written in the stream. If 'false', producer retries due to broker failures, etc., may write duplicates of the retried message in the stream. **Note that enabling idempotence requires max.in.flight.requests.per.connection to be less than or equal to 5, retries to be greater than 0 and acks must be 'all'.** If these values are not explicitly set by the user, suitable values will be chosen. If incompatible values are set, a ConfigException will be thrown.
https://docs.confluent.io/current/installation/configuration/producer-configs.html

File: src/main/java/com/Elessar/app/server/DirectMsgSender.java (hunk 1 above)
Comment (hypnotichen) [reply src/main/java/com/Elessar/app/server/DirectMsgSender.java]: why does this have anything to do with database?
all we want is to be able to distinguish direct msg sender vs kafka sender from metric log

File: src/main/java/com/Elessar/app/server/KafkaMsgSender.java [hunk 7]
```diff
@@ -40,14 +47,30 @@ public P2PMsgResponse send(List<Message> messages, String topic) throws Exceptio
...
 
-        final P2PMsgResponse.Builder p2pMsgResponse = P2PMsgResponse.newBuilder();
-        p2pMsgResponse.setSuccess(true).setIsDelivered(false);
+        /**
+         * Assumption: Messages stored in Kafka will be eventually delivered to the users by Kafka
+         * Case may fail:
+         *   Messages stored in Kafka get lost before client able to pull them down
+         *   Possible cases:
+         *      1. client crashes and not able to pull during Kafka retention period
+         *      2. Kafka cluster crash and lost all replication
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/server/KafkaMsgSender.java]: 3. client pulls the message but crashed before processing it.

File: src/main/java/com/Elessar/app/util/kafka/ProducerCreator.java [hunk 8]
```diff
@@ -13,7 +13,10 @@
     public static <K, V> Producer<K, V> create(Serializer<K> keySerializer, Serializer<V> valueSerializer) {
         Properties props = new Properties();
         props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
-        props.put(ProducerConfig.CLIENT_ID_CONFIG, "localhost:9092");
+        props.put(ProducerConfig.CLIENT_ID_CONFIG, "http://127.0.0.1:9092");
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/util/kafka/ProducerCreator.java]: ...
ok, let's add a TODO or FIXME for the IP part here. we should really use something like "ip:9000".
Please fix the port though, we should use 9000 (MyServer's port) here!

File: src/main/java/com/Elessar/app/ClientConsole.java [hunk 9]
```diff
@@ -12,48 +11,52 @@
...
 
         final MyClient client = new MyClient(serverURL, metricManager);
-        final MyClientServer clientServer = new MyClientServer("localhost", clientPort, messageQueue, metricManager);
 
-        clientServer.run();
-
-        String currUser = "";
+        if ("push".equals(mode)) {
+            final BlockingMsgQueue blockingMsgQueue = new BlockingMsgQueue();
+            final MyClientServer clientServer = new MyClientServer("localhost", clientPort, blockingMsgQueue, metricManager);
```
[the remark this answers is not in the document]
Comment (DIBL) [reply src/main/java/com/Elessar/app/ClientConsole.java]: Good point!

[the remark this answers is not in the document]
Comment (DIBL) [reply src/main/java/com/Elessar/app/ClientConsole.java]: Should we push msgQueue into myClient?

File: src/main/java/com/Elessar/app/ClientConsole.java [hunk 10]
```diff
@@ -170,6 +161,7 @@ public static void main(String[] args) {
                                     logger.info("User {} log off successfully", currUser);
                                     currUser = "";
                                     msgQueue = null;
+                                    msgQueue.close();
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/ClientConsole.java]: hehe, you would get a null pointer exception here lol

File: src/main/java/com/Elessar/app/client/MyClientServer.java [hunk 11]
```diff
@@ -25,10 +25,10 @@
     private static final Logger logger = LogManager.getLogger(MyClientServer.class);
     private final String serverName;
     private final int port;
-    private final BlockingQueue<String> messageQueue;
+    private final BlockingMsgQueue messageQueue;
     private final MetricManager metricManager;
 
-    public MyClientServer(String serverName, int port, BlockingQueue<String> messageQueue, MetricManager metricManager) {
+    public MyClientServer(String serverName, int port, BlockingMsgQueue messageQueue, MetricManager metricManager) {
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/client/MyClientServer.java]: We probably need a stop() function in this class which simply calls 
```
server.stop();
```

And in BlockingMsgQueue's close() you would call clientServer.stop()

File: src/main/java/com/Elessar/app/SetupClientMain.java [hunk 12]
```diff
@@ -16,10 +14,7 @@
      */
     public static void main(String[] args){
```
Comment (hypnotichen) [inline src/main/java/com/Elessar/app/SetupClientMain.java]: do we still need this class at all?

File: src/main/java/com/Elessar/app/ClientConsole.java (hunk 10 above)
Comment (DIBL) [reply src/main/java/com/Elessar/app/ClientConsole.java]: oops...

**Pull request merged**